### PR TITLE
[Custom Device] Clear ProcessGroup Manually

### PR DESCRIPTION
**Pull request by YanhuiDua**
### PR types
Bug fixes

### PR changes
Others 

#### Backgroud
ProcessGroupCustom creates 2 streams when init. But the NPU stream should be manually released before the DevicecContext is destroyed. 

#### Implementation
This PR reset the ProcessGroupIdMap before the default destructor of ProcessGroupIdMap. We also expand DestroyProcessGroup to all device.

1. Add `DestroyProcessGroup` function to reset ProcessGroupIdMap's value, which is ProcessGroup;
2. Expose the `DestroyProcessGroup` interface to distributed_py.cc as `ProcessGroupIdMap.destroy()`;
3. Add `destroy_process_group_id_map` API in `distributed/collective.py `. And then call `atexit.register(destroy_process_group_id_map)` in `distributed/__init__.py` to manually clear ProcessGroup

#### API changes
1. Add internal `DestroyProcessGroup(int gid)` API in [process_group.cc](https://github.com/PaddlePaddle/Paddle/pull/49182#diff-49c2a6978558fed39e3cc3723102a2f66d9334d24c52255f89fcafdd305f6136). Here, `gid` is set to 0 by default, which is same as the create function.
2. Add internal `destroy_process_group_id_map()` API in [collective.py](https://github.com/PaddlePaddle/Paddle/pull/49182#diff-578da7f9fb377a917f1cc94bf5efacd31eed06b42a00002c7ca36d1864139a97)

### CODE REVIEW

File: paddle/fluid/distributed/collective/process_group_custom.cc [hunk 1]
```diff
@@ -446,5 +446,9 @@ ProcessGroupCustom::CreateProcessGroupCustom(
   return process_group;
 }
 
+void ProcessGroupCustom::DestoryProcessGroupCustom(int gid) {
+  ProcessGroupIdMap::GetInstance()[gid].reset();
```
Comment (ronny1996) [inline paddle/fluid/distributed/collective/process_group_custom.cc]: 为什么ProcessGroupIdMap::GetInstance()[gid].reset()放在ProcessGroupCustom::DestoryProcessGroupCustom中？

Comment (YanhuiDua) [reply paddle/fluid/distributed/collective/process_group_custom.cc]: 这里是通过reset把ProcessGroupIdMap中的ProcessGroupCustom对象析构，然后把这个函数手动调用

Comment (YanhuiDua) [reply paddle/fluid/distributed/collective/process_group_custom.cc]: 这里有点问题，应该是要全部释放，不能只reset一次

Comment (YanhuiDua) [reply paddle/fluid/distributed/collective/process_group_custom.cc]: 已修改

**Review by ronny1996 — DISMISSED** [review]

File: paddle/phi/backends/stream.cc [hunk 2]
```diff
@@ -82,10 +82,13 @@ void Stream::Wait() const {
 void Stream::WaitCallback() const { callback_manager_->Wait(); }
 
 void Stream::Destroy() {
-  if (own_data_) {
+  if (own_data_ && stream_ != nullptr) {
     phi::DeviceManager::SetDevice(place_);
+    // Wait();
+    // WaitCallback();
```
Comment (qili93) [inline paddle/phi/backends/stream.cc]: 这个是冗余的注释不？还是后续有 TODO需要留着？

Comment (YanhuiDua) [reply paddle/phi/backends/stream.cc]: 是冗余注释，已修改～

**Review by qili93 — DISMISSED** [review]
LGTM

**Review by qili93 — APPROVED** [review]
LGTM

**Pull request merged**